src/Adapter/Customer/QueryHandler/GetCustomerInformationHandler.php

```diff
@@ -0,0 +1,590 @@
+            $this->getCustomerReferrers($customer),
+            $this->getCustomerAddresses($customer)
+        );
+    }
+
+    /**
+     * @param Customer $customer
+     *
+     * @return GeneralInformation
+     */
+    private function getGeneralInformation(Customer $customer)
+    {
+        return new GeneralInformation(
+            $customer->note,
+            Customer::customerExists($customer->email)
+        );
+    }
+
+    /**
+     * @param Customer $customer
+     *
+     * @return PersonalInformation
+     */
+    private function getPersonalInformation(Customer $customer)
+    {
+        $customerStats = $customer->getStats();
+
+        $gender = new Gender($customer->id_gender, $this->contextLangId);
+        $socialTitle = $gender->name ?: $this->translator->trans('Unknown', [], 'Admin.Orderscustomers.Feature');
+
+        if ($customer->birthday && '0000-00-00' !== $customer->birthday) {
+            $birthday = sprintf(
+                $this->translator->trans('%1$d years old (birth date: %2$s)', [], 'Admin.Orderscustomers.Feature'),
+                $customerStats['age'],
+                Tools::displayDate($customer->birthday)
+            );
+        } else {
+            $birthday = $this->translator->trans('Unknown', [], 'Admin.Orderscustomers.Feature');
+        }
+
+        $registrationDate = Tools::displayDate($customer->date_add, null, true);
+        $lastUpdateDate = Tools::displayDate($customer->date_upd, null, true);
+        $lastVisitDate = $customerStats['last_visit'] ?
+            Tools::displayDate($customerStats['last_visit'], null, true) :
+            $this->translator->trans('Never', [], 'Admin.Global')
+        ;
+
+        $customerShop = new Shop($customer->id_shop);
+        $customerLanguage = new Language($customer->id_lang);
+
+        $customerSubscriptions = new Subscriptions(
+            (bool) $customer->newsletter,
+            (bool) $customer->optin
+        );
+
+        return new PersonalInformation(
+            $customer->firstname,
+            $customer->lastname,
+            $customer->email,
+            $customer->isGuest(),
+            $socialTitle,
+            $birthday,
+            $registrationDate,
+            $lastUpdateDate,
+            $lastVisitDate,
+            $this->getCustomerRankBySales($customer->id),
+            $customerShop->name,
+            $customerLanguage->name,
+            $customerSubscriptions,
+            (bool) $customer->active
+        );
+    }
+
+    /**
+     * @param int $customerId
+     *
+     * @return int|null customer rank or null if customer is not ranked
+     */
+    private function getCustomerRankBySales($customerId)
+    {
+        $sql = 'SELECT SUM(total_paid_real) FROM ' . _DB_PREFIX_ . 'orders WHERE id_customer = ' . (int) $customerId . ' AND valid = 1';
+
+        if ($totalPaid = Db::getInstance()->getValue($sql)) {
+            $sql = '
+                SELECT SQL_CALC_FOUND_ROWS COUNT(*)
```

Comment:
No mistake ? `SQL_CALC_FOUND_ROWS` along with `COUNT(*)` ?

Reply:
well... i thought about trying to refactor it, but dont know if its worth, do you? it will be refactored later at some point and this code works for now.

Reply:
We need to improve this, following [the boy scout rule](https://deviq.com/boy-scout-rule/) 😉